File: nodes/equals/equals.process.py
import json

async def process(inputs, settings, config, nodeConfig):
    """
    Process function for the Equals node.
    Compares two values for equality, with deep comparison for objects and arrays.
    """
    a = inputs.get("a")
    b = inputs.get("b")
    
    # Use deep equality check
    result = deep_equals(a, b)
    
    return {"result": result}

def deep_equals(a, b):
    """Helper function for deep equality comparison"""
    # If the values are the same object, they're equal
    if a is b:
        return True
        
    # If either is None or they have different types, they're not equal
    if a is None or b is None or type(a) != type(b):
        return False
    
    # For dictionaries, compare keys and values
    if isinstance(a, dict):
        if len(a) != len(b):
            return False
        return all(k in b and deep_equals(v, b[k]) for k, v in a.items())
    
    # For lists or tuples, compare items
    if isinstance(a, (list, tuple)):
        if len(a) != len(b):
            return False
        return all(deep_equals(a[i], b[i]) for i in range(len(a)))
    
    # For other types, use standard equality
    return a == b 

File: nodes/equals/equals.process.js
export default async ({inputs, settings, config}) => {

  const a = inputs.a;
  const b = inputs.b;
  
  // Simple deep equality check
  const result = deepEquals(a, b);
  
  return { result };
};

// Helper function for deep equality comparison
function deepEquals(a, b) {
  // If the values are strictly equal, return true
  if (a === b) return true;
  
  // If either value is null or not an object, they're not equal
  if (a === null || b === null || typeof a !== 'object' || typeof b !== 'object') return false;
  
  // Get the keys of both objects
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  
  // If they have different number of keys, they're not equal
  if (keysA.length !== keysB.length) return false;
  
  // Check if all keys in A exist in B and have equal values
  return keysA.every(key => keysB.includes(key) && deepEquals(a[key], b[key]));
} 